### PR TITLE
Running masm natively

### PR DESCRIPTION
**Pull request by letmegro**
With this code I have been able to link the files and have masm run natively in the vscode terminal
currently in this version of the feature the code will only run in powershell.

### CODE REVIEW

**Review by istareatscreens — CHANGES_REQUESTED** [review]

File: extension/src/extension.ts [hunk 1]
```diff
@@ -134,8 +133,85 @@ function checkIsRunFromExplorer(fileUri: vscode.Uri): boolean {
...
+  //jwlink.exe
+  const JWLINK_EXE = vscode.Uri.joinPath(
+    native_uri_path,
+    "JWLINK",
+    "JWlink.exe"
+  );
+  //irvine lib path
+  const LIB_PATH = vscode.Uri.joinPath(native_uri_path, "irvine");
+  //irvine32.lib
+  const IRVINE32_PATH = vscode.Uri.joinPath(
```
[the remark this answers is not in the document]
Comment (letmegro) [reply extension/src/extension.ts:160]: you mean like have a function create the joinpath and return it?

File: extension/src/extension.ts [hunk 2]
```diff
@@ -134,8 +133,85 @@ function checkIsRunFromExplorer(fileUri: vscode.Uri): boolean {
...
-  terminal.sendText("echo 'Sent text immediately after creating'");
+  //create a terminal called powershell
+  if(vscode.window.terminals.length < 1){
+    vscode.window.showInformationMessage(
+      "Please ensure that a powershell terminal is open in vscode"
+    );
+  }
+  const terminal = vscode.window.terminals[0]; //integrated terminal
+  //path to the irvine to extension directory
+  let path_link: string = __dirname.slice(0, __dirname.lastIndexOf("\\"));
```
[the remark this answers is not in the document]
Comment (letmegro) [reply extension/src/extension.ts:144]: got it

File: extension/src/extension.ts [hunk 3]
```diff
@@ -134,8 +133,85 @@ function checkIsRunFromExplorer(fileUri: vscode.Uri): boolean {
...
+    temp_file.fileName.slice(0, temp_file.fileName.length - 4) +
+      "t." +
+      temp_file.fileName.slice(-3)
+  );
+  let file_data = temp_file.getText();
+  file_data = file_data.replace(
+    irvineLib32Match,
+    "INCLUDE " + path_link + IRVINE32_INC.fsPath
+  );
+  fs.writeFile(new_path.fsPath, file_data, (error) => {
```
[the remark this answers is not in the document]
Comment (letmegro) [reply extension/src/extension.ts:194]: that's doable

File: extension/src/extension.ts [hunk 4]
```diff
@@ -134,8 +133,85 @@ function checkIsRunFromExplorer(fileUri: vscode.Uri): boolean {
...
+    native_uri_path,
+    "irvine",
+    "User32.Lib"
+  );
+  const irvineLib32Match = /include.+irvine32(\.inc|)/im;
+  let temp_file = document;
+  let new_path = vscode.Uri.file(
+    temp_file.fileName.slice(0, temp_file.fileName.length - 4) +
+      "t." +
+      temp_file.fileName.slice(-3)
```
[the remark this answers is not in the document]
Comment (letmegro) [reply extension/src/extension.ts:187]: would tell the user a lot more if that was the case

File: extension/src/extension.ts [hunk 5]
```diff
@@ -134,8 +133,85 @@ function checkIsRunFromExplorer(fileUri: vscode.Uri): boolean {
...
+    } format windows pe LIBPATH ${path_link}${
+      LIB_PATH.fsPath
+    } LIBRARY ${path_link}${IRVINE32_PATH.fsPath} LIBRARY ${path_link}${
+      KERNEL32_PATH.fsPath
+    } LIBRARY ${path_link}${USER32_PATH.fsPath} file ${new_path.fsPath.slice(
+      0,
+      new_path.fsPath.length - 4
+    )}.obj ; ${new_path.fsPath.slice(0, new_path.fsPath.length - 4)}.exe`
+  );
+  //note: possibly have it instead of creating a new file replace it in the same file and change it back afterwards
```
[the remark this answers is not in the document]
Comment (letmegro) [reply extension/src/extension.ts:214]: it is not touching the user file it is touching the temp file which was created hence the new_path variable
is that still an issue?

File: extension/src/extension.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (istareatscreens) [reply extension/src/extension.ts:160]: You use closure to add the base path to the joinPath function then you just pass an array of directories and a file to the function the main function returns

File: extension/src/extension.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (istareatscreens) [reply extension/src/extension.ts:214]: I was replying to your comment on line 214

**Pull request closed without merging**